circuits/transfer/src/withdraw_from_obfuscated.rs

```diff
@@ -26,42 +39,104 @@ pub struct WithdrawFromObfuscatedCircuit {
-        let message_psk = message_ssk.public_spend_key();
+        let (input_value, input_blinder) =
+            input.decrypt(input_entropy, input_psk)?;
 
-        let message_commitment = *message.value_commitment();
-        let (message_value, message_blinder) =
-            message.decrypt(&message_r, &message_psk)?;
+        let input_commitment = *input.value_commitment();
+
+        let (change_value, change_blinder) =
+            change.decrypt(&change_entropy, change_psk)?;
+        let change_commitment = *change.value_commitment();
+
+        let identity = JubJubAffine::identity().into();
+        let change_derive_key_a = *change_psk.A();
+        let change_derive_key_b = *change_psk.B();
+
+        let (change_derive_key_public_a, change_derive_key_secret_a) =
+            if change_derive_key_is_public {
+                (change_derive_key_a, identity)
+            } else {
+                (identity, change_derive_key_a)
+            };
+
+        let (change_derive_key_public_b, change_derive_key_secret_b) =
+            if change_derive_key_is_public {
+                (change_derive_key_b, identity)
+            } else {
+                (identity, change_derive_key_b)
+            };
+
+        let change_derive_key_is_public =
+            BlsScalar::from(change_derive_key_is_public as u64);
+
+        let change_stealth_address = *change_psk
+            .gen_stealth_address(&change_entropy)
+            .pk_r()
+            .as_ref();
+
+        let change_nonce = *change.nonce();
+        let change_cipher = *change.cipher();
 
         let output_value = output.value(output_vk)?;
         let output_blinder = output.blinding_factor(output_vk)?;
         let output_commitment = *output.value_commitment();
 
         Ok(Self {
-            message_value,
-            message_blinder,
-
-            message_commitment,
-
+            input_value,
+            input_blinder,
+            input_commitment,
+            change_entropy,
+            change_value,
+            change_blinder,
+            change_commitment,
+            change_derive_key_is_public,
+            change_derive_key_public_a,
+            change_derive_key_public_b,
+            change_derive_key_secret_a,
+            change_derive_key_secret_b,
+            change_stealth_address,
+            change_nonce,
+            change_cipher,
             output_value,
             output_blinder,
             output_commitment,
         })
     }
 
     pub fn public_inputs(&self) -> Vec<PublicInputValue> {
-        // 1. commitment(Mc,mv,mb)
-        let message_commitment = JubJubAffine::from(self.message_commitment);
-
-        // 3. commitment(Oc,ov,ob)
+        let input_commitment = JubJubAffine::from(self.input_commitment);
+        let change_commitment = JubJubAffine::from(self.change_commitment);
         let output_commitment = JubJubAffine::from(self.output_commitment);
+        let change_derive_key_public_a =
```

Comment:
Delete 117 to 126 because they seem to be redundant. A from is defined [here](https://github.com/dusk-network/rusk/blob/7ddd7deaf4b93a225e8613b97ffabf45748cc76c/rusk-abi/src/public_input.rs#L47).

Reply:
`rusk-abi::PublicInputValue` exists only as a temporary replacement for `dusk-plonk::PublicInputValue` because, historically, plonk wasn't compatible with `no-std/wasm` environments.

The actual types is defined in `dusk-plonk` and didn't implement `From<JubJubExtended>`. I fixed it here: https://github.com/dusk-network/plonk/pull/574